### PR TITLE
ODE Solvers (Milestone I)

### PR DESCRIPTION
**Pull request by papachristoumarios**
This pull request considers the first milestone for my Google Summer of Code Project. 
The working fork for the project can be found [here](https://github.com/papachristoumarios/volume_approximation). 

The PR contains the following components

* ODE Solvers (located under `include/ode_solvers`).
    * Include file at `ode_solvers.h`.
    * Euler solver at `euler.h`.
    * Runge-Kutta Methods `rk.h` (generic, given by Butcher Tableau).
    * Richardson's (Bulirsch-Stoer) Extrapolation at `bulirsch_stoer.h`.
    * Leapfrog Method  at `leapfrog.h`.
    * Collocation Method (tested with polynomials for the time being). 
    * All solvers admit Neumann conditions (boundary reflections) truncated in a domain of the form `K_1 x K_2 ... K_n` where `K_i` are convex polytopes. 
    * Tests located at `tests/ode_solvers_test.cpp`.  

* Boundary (non-linear) Oracles (located under `include/nlp_oracles`)
    * Intersection of polytope with general curve of the form `sum_i a_i phi_i(t)` using interior point methods and the COIN-OR IPOPT library using the Eigen-wrapper ifopt. Can be used to find the maximum positive intersection point (in H-polytope and V-polytope) in an interval or the minimum positive intersection point (H-polytope / with higher complexity).  
    * Intersection of polytope with general curve of the form `sum_i a_i phi_i(t)` using Newton-Raphson method. 
    * Intersection of polytope with polynomial curve (or using Chebysev basis)  of the form sum_i a_i phi_i(t) using the MPSolve multi-precision polynomial solver. Can also be used with the secular basis `phi_i(t) = a_i / (t - b_i)`.  
    * Functionality is wrapped to existing functionality in `convex_bodies/{hpolytope.h, vpolytope.h}`.
    * Tests included in `tests/boundary_oracles_test.h`. 

By merging this PR 
 * functionality for implementing samplers (Hamiltonian Monte-Carlo, Langevin) is included to volesti
 * an independent ODE solver library can be used for other problems. 

Some functionality may need to be re-reviewed.

Thanks in advance,
Marios

#### Resources 

 1. Lee, Yin Tat, Ruoqi Shen, and Kevin Tian. "Logsmooth Gradient Concentration and Tighter Runtimes for Metropolized Hamiltonian Monte Carlo." arXiv preprint arXiv:2002.04121 (2020).
 2. Chevallier, Augustin, Sylvain Pion, and Frédéric Cazals. "Hamiltonian Monte Carlo with boundary reflections, and application to polytope volume calculations." (2018).
3. [MPSolve](https://github.com/robol/MPSolve)
4. [COIN-OR IPOPT](https://github.com/coin-or/Ipopt)
5. [ifopt](https://github.com/ethz-adrl/ifopt)
6. Afshar, Hadi Mohasel, and Justin Domke. "Reflection, refraction, and Hamiltonian Monte Carlo." Advances in neural information processing systems. 2015.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,7 +28,41 @@
 
 Copyright (c) 2012-2020 Vissarion Fisikopoulos  
 Copyright (c) 2018-2020 Apostolos Chalkis  
+Copyright (c) 2020-     Marios Papachristou
 
 You may redistribute or modify the software under the GNU Lesser General Public License as published by Free Software Foundation, either version 3 of the License, or (at your option) any later version. It is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY.  
 
+Main development started in 2012 by Vissarion Fisikopoulos while he was affiliated with University of Athens (UoA, Greece), and then with University of Brussels (ULB, Belgium). The main sampling and volume algorithms were developed at that time.
```
Comment (vissarion) [inline README.md]: this is the main README file of the library, I guess your edit is for the gsoc project, so please do those changes in your blog or something similar

File: include/convex_bodies/hpolytope.h [hunk 2]
```diff
@@ -16,6 +16,14 @@
 #include <Eigen/Eigen>
 #include "lp_oracles/solve_lp.h"
 
+#ifndef DISABLE_NLP_ORACLES
+#include "nlp_oracles/nlp_hpolyoracles.h"
+#include "root_finders/root_finders.h"
+#endif
+
+#define MAX_NR_TRIES 10000
```
Comment (vissarion) [inline include/convex_bodies/hpolytope.h]: this should also be inside `#ifndef DISABLE_NLP_ORACLES`

File: include/generators/v_polytopes_generators.h [hunk 3]
```diff
@@ -33,7 +37,7 @@ Polytope random_vpoly(unsigned int dim, unsigned int k, double seed = std::numer
 
     unsigned rng_seed = std::chrono::system_clock::now().time_since_epoch().count();
     RNGType rng(rng_seed);
-    if (!std::isnan(seed)) {
+    if (!isnan(seed)) {
```
Comment (vissarion) [inline include/generators/v_polytopes_generators.h:40]: why is this change needed?

File: include/nlp_oracles/nlp_hpolyoracles.h [hunk 4]
```diff
@@ -0,0 +1,301 @@
+/*
```
Comment (vissarion) [inline include/nlp_oracles/nlp_hpolyoracles.h]: please format the copyright note similarly to the other files of the library, e.g. `https://github.com/papachristoumarios/volume_approximation/blob/log-concave-sampling/include/random_walks/gaussian_cdhr_walk.hpp`

File: include/nlp_oracles/nlp_hpolyoracles.h [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+// Define the variable t we use in the optimization
+template <typename VT, typename NT>
+class HPolyOracleVariables : public VariableSet {
+public:
+  NT t, tb, eta;
+
+  HPolyOracleVariables(NT t_prev, NT tb_, NT eta_=-1): VariableSet(1, "t"), t(t_prev), tb(tb_), eta(eta_) {};
+
+  void SetVariables(const VT& T) override {
```
Comment (vissarion) [inline include/nlp_oracles/nlp_hpolyoracles.h]: is there a specific need of using virtual functions? We try to avoid them in the library.

File: include/nlp_oracles/nlp_hpolyoracles.h [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+      }
+    }
+  }
+
+
+};
+
+// Helper function that calls the optimization problem (called from hpolytope.h)
+template <typename MT, typename VT, typename Point, typename NT, class bfunc>
+std::tuple<NT, Point, int> curve_intersect_hpoly_ipopt_helper(NT t_prev, NT t0, NT eta, MT &A, VT &b, std::vector<Point> &coeffs, bfunc phi, bfunc grad_phi, std::string solution="max_pos")
```
Comment (vissarion) [inline include/nlp_oracles/nlp_hpolyoracles.h]: please wrap long lines

File: include/nlp_oracles/nlp_hpolyoracles.h [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    VectorXd phis;
+    phis.resize(M);
+
+    for (int i = 0; i < M; i++) {
+      phis(i) = (NT) (phi(t, t0, i, M));
+    }
+    return C * phis;
+  }
+
+  // Calculate jacobian matrix of constrants
```
Comment (vissarion) [inline include/nlp_oracles/nlp_hpolyoracles.h]: typo

File: include/ode_solvers/bulirsch_stoer.h [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+  funcs Fs;
+  bounds Ks;
+
+  // Contains the sub-states
+  pts xs;
+  pts xs_prev;
+
+  ptsm A;
+  bool flag;
```
Comment (vissarion) [inline include/ode_solvers/bulirsch_stoer.h]: this should be static but on top of it I find the design using flags a bit unclear and error-prone. Could you redesign by avoiding the flag?

File: include/ode_solvers/bulirsch_stoer.h [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+VolEsti is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+See the file COPYING.LESSER for the text of the GNU Lesser General
+Public License.  If you did not receive this file along with HeaDDaCHe,
+see <http://www.gnu.org/licenses/>.
+
+Implementation source: https://en.wikipedia.org/wiki/Richardson_extrapolation
```
Comment (vissarion) [inline include/ode_solvers/bulirsch_stoer.h]: Do you mean this https://en.wikipedia.org/wiki/Bulirsch%E2%80%93Stoer_algorithm ? 

Moreover, since we are using `boost` already I would like to mention that there is this library: https://www.boost.org/doc/libs/1_73_0/libs/numeric/odeint/doc/html/index.html

File: include/ode_solvers/bulirsch_stoer.h [hunk 10]
```diff
@@ -0,0 +1,224 @@
+/*
```
Comment (vissarion) [inline include/ode_solvers/bulirsch_stoer.h]: could you please use `.hpp` for header files?

File: include/convex_bodies/hpolytope.h [hunk 11]
```diff
@@ -490,6 +497,197 @@ class HPolytope{
 
     void free_them_all() {}
 
+#ifndef DISABLE_NLP_ORACLES
+
+    template <class bfunc>
+    std::tuple<NT, Point, int> curve_intersect(NT t_prev, NT t0, NT eta, std::vector<Point> &coeffs, bfunc phi, bfunc grad_phi, const std::string method="newton-raphson") {
+        if (method == "newton-raphson") {
```
Comment (vissarion) [inline include/convex_bodies/hpolytope.h]: instead of using a string and `if...then` structure you could pass directly the object of the method and call it here as it is done with random walks called by volume or sampling algorithms

File: include/ode_solvers/collocation.h [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+
+        if (ord == 0) {
+          as[i][0] = xs_prev[i];
+
+          if (exact) {
+            for (unsigned int j = 0; j < order() - 1; j++) {
+              Bs[i].row(j) = y.getCoefficients();
+            }
+          }
+
```
Comment (vissarion) [inline include/ode_solvers/collocation.h]: in general you could use less blank lines

File: include/ode_solvers/collocation.h [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+  bool precompute_flag = false;
+
+  // Function oracles x'(t) = F(x, t)
+  funcs Fs;
+
+  // Basis functions
+  bfunc phi, grad_phi;
+
+  bounds Ks;
```
Comment (vissarion) [inline include/ode_solvers/collocation.h]: Please use more declarative names for variables than `Ks, xs, zs, as` etc, it will improve readability of code a lot!

File: include/ode_solvers/rk.h [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+
+See the file COPYING.LESSER for the text of the GNU Lesser General
+Public License.  If you did not receive this file along with HeaDDaCHe,
+see <http://www.gnu.org/licenses/>.
+
+Resource: https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods
+
+*/
+
+#ifndef RK_H
```
Comment (vissarion) [inline include/ode_solvers/rk.h]: I think `runge_kutta` is a better name then `rk`

File: include/ode_solvers/rk.h [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+      Ks = bounds(xs.size(), NULL);
+      dim = xs[0].dimension();
+    };
+
+  RKODESolver(NT initial_time, NT step, pts initial_state, funcs oracles) :
+    t(initial_time), xs(initial_state), Fs(oracles), eta(step) {
+      Ks = bounds(xs.size(), NULL);
+      dim = xs[0].dimension();
+      // If no coefficients are given the RK4 method is assumed
+      cs = coeffs{0, 0.5, 0.5, 1};
```
Comment (vissarion) [inline include/ode_solvers/rk.h]: duplication of this code part could be avoided

File: reports/manuscript.tex [hunk 16]
```diff
@@ -0,0 +1,19 @@
+\documentclass[11pt]{article}
```
Comment (vissarion) [inline reports/manuscript.tex]: I would say that reports are not needed to be merged in the codebase

File: test/ode_solvers_test.cpp [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+#include <string>
+#include <typeinfo>
+
+#include "random.hpp"
+#include "random/uniform_int.hpp"
+#include "random/normal_distribution.hpp"
+#include "random/uniform_real_distribution.hpp"
+#include "random_walks/random_walks.hpp"
+
+#include "known_polytope_generators.h"
```
Comment (vissarion) [inline test/ode_solvers_test.cpp]: duplicate

File: test/ode_solvers_test.cpp [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+for more details.
+
+See the file COPYING.LESSER for the text of the GNU Lesser General
+Public License.  If you did not receive this file along with HeaDDaCHe,
+see <http://www.gnu.org/licenses/>.
+*/
+
+#include "doctest.h"
+#include "Eigen/Eigen"
+#include "ode_solvers.h"
```
Comment (vissarion) [inline test/ode_solvers_test.cpp]: please categorize include files as in other tests, i.e. first the external files with `<..>`
 and later the files of the library with `"..."`

File: test/ode_solvers_test.cpp [hunk 19]
```diff
@@ -0,0 +1,462 @@
...
+
+
+template <typename NT>
+void test_euler(){
+    typedef Cartesian<NT>    Kernel;
+    typedef typename Kernel::Point    Point;
+    typedef std::vector<Point> pts;
+    typedef std::function<Point(pts, NT)> func;
+    typedef std::vector<func> funcs;
+    typedef boost::mt19937    RNGType;
```
Comment (vissarion) [inline test/ode_solvers_test.cpp]: this is unused

Comment (vissarion) [reply test/ode_solvers_test.cpp]: the same holds for all the functions in this file

**Review by vissarion — COMMENTED** [review]
Thanks @papachristoumarios for this PR! I did a first parse mostly on design, style and typos. I will dig into it next week.

File: reports/manuscript.tex (hunk 16 above)
Comment (papachristoumarios) [reply reports/manuscript.tex]: Yes. I will remove it.

File: README.md (hunk 1 above)
Comment (papachristoumarios) [reply README.md]: OK. I left my name to the contributors only.

File: include/generators/v_polytopes_generators.h (hunk 3 above)
Comment (papachristoumarios) [reply include/generators/v_polytopes_generators.h:40]: When non-linear oracles are used (with MPSolve), `isnan` is defined as a preprocessor macro which causes errors in building the project. Hence, I just used the `using std::isnan` directive to allow it to build on multiple systems.

File: include/nlp_oracles/nlp_hpolyoracles.h (hunk 5 above)
Comment (papachristoumarios) [reply include/nlp_oracles/nlp_hpolyoracles.h]: Yes, there is need for virtual functions. The ifopt requires exactly these functions. [See example](https://github.com/ethz-adrl/ifopt/blob/master/ifopt_core/test/ifopt/test_vars_constr_cost.h)

File: include/ode_solvers/rk.h (hunk 15 above)
Comment (papachristoumarios) [reply include/ode_solvers/rk.h]: Ok. I have defined multiple constructors. Shall I remove all the constructors and use only the most basic one?

File: test/ode_solvers_test.cpp (hunk 19 above)
Comment (papachristoumarios) [reply test/ode_solvers_test.cpp]: Thank you. I had v-polys in the previous version of the tests w/ the previous versions of the project. 
I removed the redundant typedefs.

File: include/nlp_oracles/nlp_hpolyoracles.h (hunk 4 above)
Comment (papachristoumarios) [reply include/nlp_oracles/nlp_hpolyoracles.h]: Ok. I will convert the format to match the correct one.

File: include/nlp_oracles/nlp_hpolyoracles.h (hunk 6 above)
Comment (papachristoumarios) [reply include/nlp_oracles/nlp_hpolyoracles.h]: fixed

File: include/ode_solvers/collocation.h (hunk 13 above)
Comment (papachristoumarios) [reply include/ode_solvers/collocation.h]: Most of them follow the mathematical notation. We can change them to be more descriptive.
I personally prefer using the math symbols as they are when writing code that is correlated to maths directly,
since one can look at the papers and find out about the code directly. From a sweng perspective, I acknowledge
that it is not the best practice.

File: test/ode_solvers_test.cpp (hunk 17 above)
Comment (papachristoumarios) [reply test/ode_solvers_test.cpp]: removed

Comment (papachristoumarios) [reply test/ode_solvers_test.cpp]: solved.

File: include/ode_solvers/bulirsch_stoer.h (hunk 10 above)
Comment (papachristoumarios) [reply include/ode_solvers/bulirsch_stoer.h]: Everything renamed to `.hpp`

File: include/ode_solvers/bulirsch_stoer.h (hunk 9 above)
Comment (papachristoumarios) [reply include/ode_solvers/bulirsch_stoer.h]: Mea cupla. It is the richardon extrapolation solver. I renamed it.

File: include/convex_bodies/hpolytope.h (hunk 11 above)
Comment (papachristoumarios) [reply include/convex_bodies/hpolytope.h]: Could you please locate a specific file and line to adjust the design pattern accordingly? 

AFAIU, you mean a pointer to a member function (according to which we do the `curve_intersect`). [See here also](https://stackoverflow.com/questions/8865766/get-a-pointer-to-objects-member-function)

File: include/ode_solvers/rk.h (hunk 15 above)
Comment (vissarion) [reply include/ode_solvers/rk.h]: Yes, please use only the ones you need.

File: include/ode_solvers/collocation.h (hunk 13 above)
Comment (vissarion) [reply include/ode_solvers/collocation.h]: ok, I think I prefer the mathematical notation, so if you agree please add some comment such as the variable names use the math notation from paper [..]

File: include/convex_bodies/hpolytope.h (hunk 11 above)
Comment (vissarion) [reply include/convex_bodies/hpolytope.h]: I have this example in mind. The volume algorithm as instantiated here: https://github.com/GeomScale/volume_approximation/blob/develop/test/volume_cb_hpolytope.cpp#L65 is using the `CDHRWalk` defined here https://github.com/GeomScale/volume_approximation/blob/develop/include/random_walks/uniform_cdhr_walk.hpp#L23
Of course you may want to pass a reference to an object e.g. NewtonRaphson that has an apply method to do the computation.

File: include/ode_solvers/rk.h (hunk 15 above)
Comment (papachristoumarios) [reply include/ode_solvers/rk.h]: Done.

File: include/ode_solvers/collocation.h (hunk 13 above)
Comment (papachristoumarios) [reply include/ode_solvers/collocation.h]: Ok I added a general description on the main header file `include/ode_solvers/ode_solvers.hpp`.

File: include/convex_bodies/hpolytope.h (hunk 11 above)
Comment (papachristoumarios) [reply include/convex_bodies/hpolytope.h]: Eventually, I have kept the non-linear boundaries as members of the polytope classes and added the macro `DISABLE_NLP_ORACLES` to exclude them from the compilation, if desired. 
Another thing I had in mind when I decided to use the macros is keeping the architecture consistent since
`line_intersect_*` methods are member functions. 
The helper functions can indeed be called using their pointers but I do not see any point in doing this. 

Shall I move everything out, eventually, and do as you propose?

File: include/ode_solvers/bulirsch_stoer.h (hunk 8 above)
Comment (papachristoumarios) [reply include/ode_solvers/bulirsch_stoer.h]: Done.

File: include/convex_bodies/hpolytope.h (hunk 11 above)
Comment (papachristoumarios) [reply include/convex_bodies/hpolytope.h]: NLP oracles refactored.

**Comment by papachristoumarios** [comment]
@vissarion All the recommendations have been addressed.

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 20]
```diff
@@ -1,38 +0,0 @@
----
```
Comment (vissarion) [inline .github/ISSUE_TEMPLATE/bug_report.md]: is this deleted by accident ?

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 21]
```diff
@@ -1,20 +0,0 @@
----
```
Comment (vissarion) [inline .github/ISSUE_TEMPLATE/feature_request.md]: is this deleted by accident ?

File: README.md [hunk 22]
```diff
@@ -28,7 +28,36 @@
...
 
+Later, Chalkis Apostolos affiliated with University of Athens (UoA, Greece) and mentored by V.F. and partially supported by GSoC'2018, 2019 programs enhanced volesti with more algorithms for sampling and volume as well as an `R` interface.       
+
+We acknowledge several contributions by the open-source community, most notably a `python` interface by Pedro Zuidberg Dos Martires affiliated with KU Leuven.
+
+#### Project Details
```
Comment (vissarion) [inline README.md]: notice that this is the main README of the whole library!

File: include/convex_bodies/vpolytope.h [hunk 23]
```diff
@@ -449,6 +447,19 @@ class VPolytope{
...
 
+        template <class bfunc, class NonLinearOracle>
+        std::tuple<NT, Point, int> curve_intersect(
+          NT t_prev,
+          NT t0,
+          NT eta,
+          std::vector<Point> &coeffs,
+          bfunc phi,
+          bfunc grad_phi,
+          NonLinearOracle &oracle,
```
Comment (vissarion) [inline include/convex_bodies/vpolytope.h]: maybe `intersection_oracle` or `boundary_oracle` are more descriptive variable names

File: include/nlp_oracles/nlp_hpolyoracles.hpp [hunk 24]
```diff
@@ -0,0 +1,557 @@
+/*
```
Comment (vissarion) [inline include/nlp_oracles/nlp_hpolyoracles.hpp]: please use copyright comments consistently

File: include/nlp_oracles/nlp_vpolyoracles.hpp [hunk 25]
```diff
@@ -0,0 +1,319 @@
+/*
```
Comment (vissarion) [inline include/nlp_oracles/nlp_vpolyoracles.hpp]: please use copyright comments consistently

File: include/root_finders/mp_solve_wrapper.hpp [hunk 26]
```diff
@@ -0,0 +1,92 @@
+/*
```
Comment (vissarion) [inline include/root_finders/mp_solve_wrapper.hpp]: please use copyright comments consistently

File: include/root_finders/newton_raphson.hpp [hunk 27]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (vissarion) [inline include/root_finders/newton_raphson.hpp]: please use copyright comments consistently

File: test/CMakeLists.txt [hunk 28]
```diff
@@ -11,6 +11,38 @@ CMAKE_MINIMUM_REQUIRED(VERSION 2.4.5)
 
 set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS true)
 
+option(DISABLE_NLP_ORACLES "Disable non-linear oracles (used in collocation)" OFF)
```
Comment (vissarion) [inline test/CMakeLists.txt]: I think it is better for now the default value to be `ON`

**Review by vissarion — COMMENTED** [review]
Thanks for the changes! I have a few more comments.
Also, please,
* remove  `reports/references.bib`
* use copyright comments consistently

**Comment by papachristoumarios** [comment]
> Thanks for the changes! I have a few more comments.
> Also, please,
> 
> * remove  `reports/references.bib`
> * use copyright comments consistently

@vissarion All requests have been resolved.

File: include/ode_solvers/collocation.hpp [hunk 29]
```diff
@@ -19,7 +19,7 @@ template <
   typename NT,
   class Polytope,
   class bfunc,
-  class func=std::function <Point(std::vector<Point>, NT)>,
+  class func=std::function <Point(std::vector<Point>&, NT&)>,
```
Comment (elias-tsigaridas) [inline include/ode_solvers/collocation.hpp:22]: IMO we should  use functors instead of a pointer to function. 
This, besides speeding up the code, will make the code more clear.

This comment applies to all the ODE solvers

File: include/ode_solvers/euler.hpp [hunk 30]
```diff
@@ -11,7 +11,7 @@
 #ifndef EULER_HPP
 #define EULER_HPP
 
-template <typename Point, typename NT, class Polytope, class func=std::function <Point(std::vector<Point>, NT)>>
+template <typename Point, typename NT, class Polytope, class func=std::function <Point(std::vector<Point>&, NT&)>>
 class EulerODESolver {
```
Comment (elias-tsigaridas) [inline include/ode_solvers/euler.hpp:15]: As a general principle, I do not like the use of pointers. In this context, I think that we should consider another implementation of the bounds. 
We expect that Polytope is a big object and thus we should rely on move semantics and not on pointers to handle it.  
Also, I think that the size of bounds should be known in compile time.

**Review by vissarion — APPROVED** [review]

**Comment by vissarion** [comment]
I approved this PR. Thanks @papachristoumarios once more!
@TolisChal @elias-tsigaridas are you ok with merging (after passing the tests successfully)?

**Review by TolisChal — APPROVED** [review]

**Review by elias-tsigaridas — APPROVED** [review]
